clang/include/clang/Serialization/ASTReader.h

```diff
@@ -2220,33 +2221,40 @@ class ASTReader
     return Sema::AlignPackInfo::getFromRawEncoding(Raw);
   }
 
+  using RawLocEncoding = SourceLocationEncoding::RawLocEncoding;
+
   /// Read a source location from raw form and return it in its
   /// originating module file's source location space.
-  SourceLocation ReadUntranslatedSourceLocation(SourceLocation::UIntTy Raw,
-                                                LocSeq *Seq = nullptr) const {
+  std::pair<SourceLocation, unsigned>
+  ReadUntranslatedSourceLocation(RawLocEncoding Raw,
+                                 LocSeq *Seq = nullptr) const {
     return SourceLocationEncoding::decode(Raw, Seq);
   }
 
   /// Read a source location from raw form.
-  SourceLocation ReadSourceLocation(ModuleFile &ModuleFile,
-                                    SourceLocation::UIntTy Raw,
-                                    LocSeq *Seq = nullptr) const {
-    SourceLocation Loc = ReadUntranslatedSourceLocation(Raw, Seq);
-    return TranslateSourceLocation(ModuleFile, Loc);
+  SourceLocation ReadSourceLocation(ModuleFile &MF, RawLocEncoding Raw,
+                                       LocSeq *Seq = nullptr) const {
+    if (!MF.ModuleOffsetMap.empty())
+      ReadModuleOffsetMap(MF);
+
+    auto [Loc, ModuleFileIndex] = ReadUntranslatedSourceLocation(Raw, Seq);
+    ModuleFile *OwningModuleFile =
+        ModuleFileIndex ? MF.DependentModules[ModuleFileIndex - 1] : &MF;
+    return TranslateSourceLocation(*OwningModuleFile, Loc);
   }
 
   /// Translate a source location from another module file's source
   /// location space into ours.
   SourceLocation TranslateSourceLocation(ModuleFile &ModuleFile,
                                          SourceLocation Loc) const {
-    if (!ModuleFile.ModuleOffsetMap.empty())
-      ReadModuleOffsetMap(ModuleFile);
-    assert(ModuleFile.SLocRemap.find(Loc.getOffset()) !=
-               ModuleFile.SLocRemap.end() &&
-           "Cannot find offset to remap.");
-    SourceLocation::IntTy Remap =
-        ModuleFile.SLocRemap.find(Loc.getOffset())->second;
-    return Loc.getLocWithOffset(Remap);
+    if (Loc.isInvalid())
+      return Loc;
+
+    // It implies that the Loc is already translated.
+    if (SourceMgr.isLoadedSourceLocation(Loc))
+      return Loc;
```

Comment:
How can we get here with a `SourceLocation` that's already been translated?

Reply:
My thought was like "all the recorded source location are local to its module file", so it won't be translated. But after I review how we judge whether a source location is loaded in source manager, I find this may be dangerous in extreme cases.  We judge if a source location is loaded by comparing the offset with the current loaded offset. So it may be dangerous if we have too many loaded source location in the TU and the source location we read is too big, then we may misread it as loaded source location.

My intention for this code is to make `TranslateSourceLocation` to be symmetric with the original one. The original `TranslateSourceLocation` is re-enterable. We could call `TranslateSourceLocation`  on a translated source location. But we can't do this for the `TranslateSourceLocation`   in the current patch.

So I just removed the check and leave a FIXME comment. I'd like to fix this as a NFC patch after we land this.

Reply:
I don't think this an issue. Now that `TranslateSourceLocation()` is only called from `ReadSourceLocation()` there's no reason for taking care to handle already-translated source locations correctly. Maybe now that it doesn't use any `ASTWriter` members we can make the function a non-member and make it static `ASTWriter.cpp`. In my mind that's enough to leave out the FIXME and just assume it only works when used on an untranslated source location read in `ReadSourceLocation()`.

As an optional clarification we can return a special type (e.g. `UntranslatedSourceLocation`) from `ReadUntranslatedSourceLocation()` and make it so that only `TranslateSourceLocation()` can transform that into the regular (now always-translated) `SourceLocation`.

Reply:
> Now that TranslateSourceLocation() is only called from ReadSourceLocation()

Sadly, this is not true. `TranslateSourceLocation()` may be called in `ASTReader::ReadAST()`:

https://github.com/llvm/llvm-project/blob/aac4d03423dd6b7bdef0f2eb03c570f3e2ca6630/clang/lib/Serialization/ASTReader.cpp#L4588-L4591

The input value of  `TranslateSourceLocation()` there may come from a reading of untranslated source location in `ASTReader::ReadControlBlock` when reading imported modules. Or the input value may come from the argument of  `ASTReader::ReadAST()`, where must be a translated source location. 

Then it looks really dangerous to me. So I add the FIXME. We may not be able to change the signature of the argument of `TranslateSourceLocation()` to `UntranslatedSourceLocation` since that will require us to change the signature of `ASTReader::ReadAST()`.

The reason why actually it works, is that, in the case the input value comes from a translated source location (passed directly in `ASTReader::ReadAST()`), the value of `M.ImportedBy` may always be null **now**.  But I feel it is dangerous if someone changes it suddenly.

Reply:
I see, thanks for explaining that.

I think we should still attempt to unify things so that `TranslateSourceLocation()` is only called from `ReadSourceLocation()`. Even if it means carrying something like `std::variant<SourceLocation, SourceLocationEncoding::RawLocEncoding>` (or `, UntranslatedSourceLocation>`) in few places. Juggling translated and untranslated `SourceLocation` instances and making functions resilient to both seems unnecessarily complex. I'm fine with leaving a FIXME in this patch and doing a follow-up.